### PR TITLE
Add new interfaces to improve data consuming performance. (#65)

### LINKED ISSUE #65: [Improve] Add interfaces in `class Message` to avoid unnecessary memory copy.
## Motivation

Currently `class Message` provides two groups of functions to access the message's payload:

1. `Message::getData` and `Message::getLength`.
2. `getDataAsString`: this is a wrapper of the two above.

The second one performs `std::string`'s construction, which contains expensive memory copy. If `class Message` can provide a new function that gives payload's ownership to the caller like this:

```cpp
// In SharedBuffer.h add new function:
bool SharedBuffer::pop(std::string &target) {
    // Swap *data_ into target, then reset ptr_, readIdx_ etc.
}

// Message.cc
void Message::moveDataIntoString(std::string &data) {
    // Move payload into data via SharedBuffer::pop.
}
```

In high-throughput scenario, these new functions could bring significant performance improvement at data consuming.

### PR DESCRIPTION
**Pull request by tongsucn**
Fixes #65

### Motivation

Improve data consuming performance by adding new interfaces on `class Message` and `class SharedBuffer`.

### Modifications

Adding two functions:

`lib/SharedBuffer.h`

```cpp
void SharedBuffer::pop(std::string &target)
```

`pulsar/Message.h`, `lib/Message.cc`

```cpp
std::string Message::release()
```

### Verifying this change

This change added tests and can be verified as follows:

Run client's unit test, it's added in `MessageTest.cc`.

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [ ] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [x] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Review by BewareMyPower — COMMENTED** [review]
How about returning a rvalue reference? e.g.

```c++
std::string&& releaseData();
```

Though the output parameter is widely used in Pulsar C++ client, it's because the returned value is `Result`. In this case, using an output parameter looks ugly IMO.

File: lib/SharedBuffer.h [hunk 1]
```diff
@@ -203,6 +203,13 @@ class SharedBuffer {
         writeIdx_ = 0;
     }
 
+    bool pop(std::string& target) {
+        if (data_ == nullptr) return false;
```
Comment (BewareMyPower) [inline lib/SharedBuffer.h:207]: `data_` could never be a `nullptr`

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: lib/SharedBuffer.h (hunk 1 above)
Comment (BewareMyPower) [inline lib/SharedBuffer.h:207]: The returned value is always ignored. I think we should throw an exception to disable `moveDataIntoString` on an invalid `Message` object, i.e. built from a `MessageBuilder` whose content was set by `setAllocatedContent` .

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

**Review by RobertIndie — CHANGES_REQUESTED** [review]
Please add some unit tests.

File: lib/SharedBuffer.h (hunk 1 above)
Comment (tongsucn) [reply lib/SharedBuffer.h:209]: Good point. Let me add a flag in `SharedBuffer` to denote if it's a read-only one.

Comment (tongsucn) [reply lib/SharedBuffer.h:209]: Good point.

Comment (tongsucn) [reply lib/SharedBuffer.h]: Done

**Comment by shibd** [comment]
@tongsucn Please rebase the main branch.

**Comment by tongsucn** [comment]
> Please add some unit tests.

Done

File: include/pulsar/Message.h [hunk 2]
```diff
@@ -100,6 +100,20 @@ class PULSAR_PUBLIC Message {
...
+     * @return the released payload.
+     */
+    std::string releaseData();
+
+    /**
+     * Move payload into the given string.
+     *
+     * @param data the string that receiving the payload's ownership.
+     */
+    void moveDataIntoString(std::string& data);
```
Comment (shibd) [inline include/pulsar/Message.h:115]: useless interface.

Comment (tongsucn) [reply include/pulsar/Message.h]: Done

**Review by shibd — APPROVED** [review]

File: tests/MessageTest.cc [hunk 3]
```diff
@@ -53,6 +53,15 @@ TEST(MessageTest, testAllocatedContents) {
     Message msg = msgBuilder.build();
     ASSERT_FALSE(strncmp("content", (char*)msg.getData(), msg.getLength()));
     ASSERT_EQ(content, (char*)msg.getData());
+
+    bool throwed = false;
+    try {
+        msg.releaseData();
+    } catch (std::runtime_error& ex) {
```
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply tests/MessageTest.cc:60]: +1

File: tests/MessageTest.cc [hunk 4]
```diff
@@ -53,6 +53,8 @@ TEST(MessageTest, testAllocatedContents) {
     Message msg = msgBuilder.build();
     ASSERT_FALSE(strncmp("content", (char*)msg.getData(), msg.getLength()));
     ASSERT_EQ(content, (char*)msg.getData());
+
+    ASSERT_THROW(msg.releaseData(), std::runtime_error);
```
Comment (RobertIndie) [inline tests/MessageTest.cc:57]: Please add a test case for releasing the data successfully and verify the data is actually swapped.

**Review by RobertIndie — CHANGES_REQUESTED** [review]

**Comment by tongsucn** [comment]
@BewareMyPower @shibd @RobertIndie I found a problem on the new interface:

When batching is enabled on producer side, `Consumer::receive` actually creates a shallow copy of the whole message batch and sets reading offset (i.e. `SharedBuffer::readIdx_`) via `SharedBuffer::slice` in `Commands::deSerializeSingleMessageInBatch`.

It means that the messages from a batch cannot be **_moved_** into another `std::string` without exposing the reading offset. Therefore I think the current implementation of `Message::releaseData` or `Message::releaseData` itself is not a good design. If you have any better ideas, please tell me, or just close the issue and PR.

**Comment by BewareMyPower** [comment]
@tongsucn Good point. I think it's determined by how do you use that.

If you only want a view of the Message's value, you can create an instance `std::string_view` (or a similar class). But you should be careful that the lifetime is the same as `Message`, i.e. you must keep the `Message` objects in memory to make `std::string_view` valid.

If you want some extra in place operations, I think you can expose the non-const `data()` override.

```c++
const void* getData() const;
void* getData();  // new interface
```

If your process function in application just accepts a `std::string`, I'm afraid you have to change it to `std::string_view`.

**Comment by BewareMyPower** [comment]
I just found the `batchReceive` method added in #21. Maybe if the producer enables batching, the consumer must combine `batchReceive` and the `releaseData` APIs. In this case, I prefer not throwing an exception when `data_` is `nullptr`. Instead, fallback to copying the data rather than moving.

**Comment by tongsucn** [comment]
> I just found the `batchReceive` method added in #21. Maybe if the producer enables batching, the consumer must combine `batchReceive` and the `releaseData` APIs. In this case, I prefer not throwing an exception when `data_` is `nullptr`. Instead, fallback to copying the data rather than moving.

Well, I think the result of `batchReceive` may contain data from multiple message batches, which make the situation more complicated. Falling back without telling caller may confuse the callers.

  Under current design, `pulsar::Message` is actually an *agent* of the message data rather than a *container*. An *agent* should NOT give the ownership of the data to others, because it actually **DOES NOT OWN** the data. Therefore, I think interfaces like `releaseData` should be avoided.

**Pull request closed without merging**